src/Microsoft.AspNetCore.Authentication2.Cookies/CookieAuthenticationHandler.cs

```diff
@@ -0,0 +1,461 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+
+using System;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.Authentication2.Cookies
+{
+    public class CookieAuthenticationHandler : AuthenticationSchemeHandler<CookieAuthenticationOptions>
+    {
+        private const string HeaderValueNoCache = "no-cache";
+        private const string HeaderValueMinusOne = "-1";
+        private const string SessionIdClaim = "Microsoft.AspNetCore.Authentication2.Cookies-SessionId";
+
+        private bool _finishCalled;
+        private bool _shouldRefresh;
+        private DateTimeOffset? _refreshIssuedUtc;
+        private DateTimeOffset? _refreshExpiresUtc;
+        private string _sessionKey;
+        private Task<AuthenticateResult> _readCookieTask;
+
+        public async override Task InitializeAsync(AuthenticationScheme scheme, HttpContext context)
+        {
+            await base.InitializeAsync(scheme, context);
+
+            // TODO: This needs to go into some kind of base class for reuse
+            if (Options.EventsType != null)
+            {
+                Options.Events = context.RequestServices.GetRequiredService(Options.EventsType) as CookieAuthenticationEvents;
+            }
+
+            if (Options.Events == null)
+            {
+                Options.Events = new CookieAuthenticationEvents();
+            }
+            if (String.IsNullOrEmpty(Options.CookieName))
+            {
+                Options.CookieName = CookieAuthenticationDefaults.CookiePrefix + Scheme.Name;
```

Comment:
The handlers are initialized per request, as enabling DI activation is a goal now.  So each auth options is not really like our IOptions instances, each is actually per named scheme instance, similar to what we have today, and why I stuff it into the dictionary of each scheme...  (RIP named options)